File: ml-benchmark/autolifter/single-pass/mps.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

/*User provided programs*/

mps = fix (lambda f: List -> Int. lambda l: List.
  match l with
    nil _ -> 0
  | cons {h, t} ->
    let res = (f t) in
      if < 0 (+ res h) then 0 else + res h
  end
);

main = single_pass mps;

File: ml-benchmark/autolifter/single-pass/mps.ml
type list_ =
  | Cons of int * list_
  | Nil of unit

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let rec mps l =
  match l with
  | Nil _ -> 0
  | Cons (h, t) ->
      let res = mps t in
      if 0 < res + h then 0 else res + h

let program = single_pass mps